src/apps/js_v8/tmpl/receipt.cpp

```diff
@@ -0,0 +1,128 @@
+#include "receipt.h"
+
+#include "ccf/receipt.h"
+#include "ds/logger.h"
+#include "template.h"
+
+namespace ccf::v8_tmpl
+{
+  static ccf::Receipt* unwrap_receipt(v8::Local<v8::Object> obj)
+  {
+    return static_cast<ccf::Receipt*>(
+      obj->GetAlignedPointerFromInternalField(0));
+  }
+
+  static void get_signature(
+    v8::Local<v8::Name> name, const v8::PropertyCallbackInfo<v8::Value>& info)
+  {
+    ccf::Receipt* receipt = unwrap_receipt(info.Holder());
+    v8::Local<v8::String> value =
+      v8_util::to_v8_str(info.GetIsolate(), receipt->signature);
+    info.GetReturnValue().Set(value);
+  }
+
+  static void get_node_cert(
+    v8::Local<v8::Name> name, const v8::PropertyCallbackInfo<v8::Value>& info)
+  {
+    ccf::Receipt* receipt = unwrap_receipt(info.Holder());
+    v8::Local<v8::Value> value;
+    if (receipt->cert.has_value())
+      value = v8_util::to_v8_str(info.GetIsolate(), receipt->cert.value());
+    else
+      value = v8::Undefined(info.GetIsolate());
+    info.GetReturnValue().Set(value);
+  }
+
+  static void get_leaf(
+    v8::Local<v8::Name> name, const v8::PropertyCallbackInfo<v8::Value>& info)
+  {
+    ccf::Receipt* receipt = unwrap_receipt(info.Holder());
+    v8::Local<v8::String> value =
+      v8_util::to_v8_str(info.GetIsolate(), receipt->leaf);
+    info.GetReturnValue().Set(value);
+  }
+
+  static void get_node_id(
+    v8::Local<v8::Name> name, const v8::PropertyCallbackInfo<v8::Value>& info)
+  {
+    ccf::Receipt* receipt = unwrap_receipt(info.Holder());
+    v8::Local<v8::String> value =
+      v8_util::to_v8_str(info.GetIsolate(), receipt->node_id.value());
+    info.GetReturnValue().Set(value);
+  }
+
+  static void get_proof(
+    v8::Local<v8::Name> name, const v8::PropertyCallbackInfo<v8::Value>& info)
+  {
+    v8::Isolate* isolate = info.GetIsolate();
+    v8::Local<v8::Context> context = isolate->GetCurrentContext();
+    ccf::Receipt* receipt = unwrap_receipt(info.Holder());
+
+    size_t size = receipt->proof.size();
+    std::vector<v8::Local<v8::Value>> elements;
+    elements.reserve(size);
+    uint32_t i = 0;
+    for (auto& element : receipt->proof)
+    {
+      auto is_left = element.left.has_value();
+      v8::Local<v8::Object> obj = v8::Object::New(isolate);
+      obj
+        ->Set(
+          context,
+          v8_util::to_v8_istr(isolate, is_left ? "left" : "right"),
+          v8_util::to_v8_str(
+            isolate, (is_left ? element.left : element.right).value()))
+        .Check();
+      elements.push_back(obj);
+    }
+
+    v8::Local<v8::Array> array =
+      v8::Array::New(info.GetIsolate(), elements.data(), size);
+
+    info.GetReturnValue().Set(array);
+  }
+
+  v8::Local<v8::ObjectTemplate> Receipt::create_template(v8::Isolate* isolate)
+  {
+    v8::EscapableHandleScope handle_scope(isolate);
+
+    v8::Local<v8::ObjectTemplate> tmpl = v8::ObjectTemplate::New(isolate);
+
+    // Field 0: ccf::Receipt
+    tmpl->SetInternalFieldCount(1);
```

Comment:
Better if named, as mentioned above. Could also be wrapped into a call that takes an enum for safety.